src/txn/txn.c

```diff
@@ -207,12 +210,20 @@ __txn_get_snapshot_int(WT_SESSION_IMPL *session, bool publish)
     n = 0;
 
     /* Fast path if we already have the current snapshot. */
-    if ((commit_gen = __wt_session_gen(session, WT_GEN_COMMIT)) != 0) {
-        if (F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) && commit_gen == __wt_gen(session, WT_GEN_COMMIT))
+    if ((snapshot_gen = __wt_session_gen(session, WT_GEN_HAS_SNAPSHOT)) != 0) {
+        WT_ASSERT(
+          session, F_ISSET(txn, WT_TXN_HAS_SNAPSHOT) || !F_ISSET(txn, WT_TXN_REFRESH_SNAPSHOT));
+        if (!F_ISSET(txn, WT_TXN_REFRESH_SNAPSHOT) &&
+          snapshot_gen == __wt_gen(session, WT_GEN_HAS_SNAPSHOT))
             return;
-        __wt_session_gen_leave(session, WT_GEN_COMMIT);
+        /*
+         * Leave the generation here and enter again later to acquire a new snapshot if any
+         * concurrent transactions have been committed and changed the global snapshot generation
+         * after we entered the generation.
```

Comment:
I'm not understanding this comment. I believe it's safe to release this generation here, because the only thing the generation is used for is checking whether a thread already has a valid snapshot.
If the generation is used for tracking and validating global state (as the comment implies) releasing wouldn't be safe.

Reply:
Fixed in ab154f8f95999594224485f115c27d2b986d5cf4